### DISCUSSION TITLE
Using Cookbooks to Upgrade an Environment Not Deployed Using an Earlier Version of the Template

### DISCUSSION THREAD

**Discussion by td-gis** [Esri/arcgis-cookbook#398] [General]
Hello,

I was hoping to garner more support and adoption for chef tooling within an organization to manage their ArcGIS Enterprise environments. A potential roadblock I foresee is the stipulation that the arcgis-enterprise-base\readme includes, recommending not to administer or upgrade environments not initially deployed using an earlier version of the template. 

Ideally I'd like to understand why this recommendation is in place. 

Referencing the arcgis-5.0.0-cookbooks\templates\arcgis-enterprise-base\README.

**Comment by cameronkroeker** [comment]
Hi @td-gis,

This is because there are many combinations of recipes/attributes/settings which can lead to unknown or unforeseen outcomes which is not ideal when it comes to upgrades. It can lead to a "trial and error" approach which is fine for fresh deployments because these can be torn down and started over. With upgrades if it fails then there will be downtime introduced in troubleshooting the cause, or even possibly having to revert to snapshots/backups each time when trying to correct what went wrong. If the risks are understood then this is an approach that can be taken, however this is why it is not recommended.

When a deployment is created using a previous version of one of the templates and then upgraded with a new version of that template we have a high confidence and certainty that it will succeed because this approach has been tested and certified. 

If you have deployed with Chef not using one of the templates, then its recommended to use/update the json files originally used for the upgrades. Or if the environment was originally deployed manually (or via some other means), then it would be recommended to stand up a new environment along side it using Chef, migrate all the content over to it, then once that new environment has been verified to work the old environment can be decommissioned. Then moving forward the new environment can be upgraded using Chef. 

Thanks,
Cameron K.

**Reply by td-gis** [reply]
Hi Cameron, 

Thanks for the response! That was my suspicion.. I'd considered using wegisdr to restore a deployment to one installed with templates, but that would be a bit cumbersome for larger organizations. I may look into the troubleshooting required on lower environments. 

Thanks!
